lib/compat/wordpress-6.8/blocks.php

```diff
@@ -20,77 +20,121 @@ function gutenberg_stabilize_experimental_block_supports( $args ) {
 		return $args;
 	}
 
-	$experimental_to_stable_keys = array(
-		'typography'           => array(
+	$experimental_supports_map       = array( '__experimentalBorder' => 'border' );
+	$common_experimental_properties  = array(
+		'__experimentalDefaultControls'   => 'defaultControls',
+		'__experimentalSkipSerialization' => 'skipSerialization',
+	);
+	$experimental_support_properties = array(
+		'typography' => array(
 			'__experimentalFontFamily'     => 'fontFamily',
 			'__experimentalFontStyle'      => 'fontStyle',
 			'__experimentalFontWeight'     => 'fontWeight',
 			'__experimentalLetterSpacing'  => 'letterSpacing',
 			'__experimentalTextDecoration' => 'textDecoration',
 			'__experimentalTextTransform'  => 'textTransform',
 		),
-		'__experimentalBorder' => 'border',
 	);
+	$done                            = array();
 
 	$updated_supports = array();
 	foreach ( $args['supports'] as $support => $config ) {
-		// Add the support's config as is when it's not in need of stabilization.
-		if ( empty( $experimental_to_stable_keys[ $support ] ) ) {
+		/*
+		 * If this support config has already been stabilized, skip it.
+		 * A stable support key occurring after an experimental key, gets
+		 * stabilized then so that the two configs can be merged effectively.
+		 */
```

Comment:
> Is it possible for the same support to be declared twice for the same block?

Yes. Plugins can filter block type args. 

If core blocks use the stable config but a plugin still filters block type args using the experimental flags we'll get both. Likewise, if a block still uses the experimental key and a plugin filters using the stable key, we'll get both again.

In the case that we have an experimental key based config when it gets stabilized we need to see if it should be merged with any other stable config (or perhaps use the stable config value if it isn't an array/object). So if the experimental key is reached before the stable, we need to stabilize the config under the stable key while processing the experimental config so that the stable merged result can be stored under the stable key without overwriting anything that might have been filtered separately there.

As the loop continues and we reach a stable config key, if it was already stabilized and merged when processing the earlier experimental config, there's no need to do it again.

Does that make sense?

Reply:
That does make sense yes! I wasn't considering that plugins will be using the exact same filter we're using here 😄

I'm guessing that when this goes in core it will become a method of `class WP_Block_Type` instead?

Reply:
> I'm guessing that when this goes in core it will become a method of class WP_Block_Type instead?

That's correct and touched on in this [comment](https://github.com/WordPress/gutenberg/pull/67018#discussion_r1847969189) in this PR. There's also some more context in the original [typography support stabilization PR](https://github.com/WordPress/gutenberg/pull/63401) and it's [backport](https://github.com/WordPress/wordpress-develop/pull/7069). 

That [backport](https://github.com/WordPress/wordpress-develop/pull/7069) though is out of date and needs updating with the other recent stabilizations.